data_pes.go

```diff
@@ -412,3 +419,329 @@ func parseESCR(i *astikit.BytesIterator) (cr *ClockReference, err error) {
+		if err != nil {
+			return 0, err
+		}
+		bytesWritten += n
+	}
+
+	if h.HasAdditionalCopyInfo {
+		b.Write(true) // marker_bit
+		b.WriteN(h.AdditionalCopyInfo, 7)
+		bytesWritten++
+	}
+
+	if h.HasCRC {
+		// TODO, not supported
+	}
+
+	if h.HasExtension {
+		// exp 10110001
+		// act 10111111
+		b.Write(h.HasPrivateData)
+		b.Write(false) // TODO pack_header_field_flag, not implemented
+		//b.Write(h.HasPackHeaderField)
+		b.Write(h.HasProgramPacketSequenceCounter)
+		b.Write(h.HasPSTDBuffer)
+		b.WriteN(uint8(0xff), 3) // reserved
+		b.Write(h.HasExtension2)
+		bytesWritten++
+
+		if h.HasPrivateData {
+			writeBytesN(&b, h.PrivateData, 16, 0)
+			bytesWritten += 16
+		}
+
+		if h.HasPackHeaderField {
+			// TODO (see parsePESOptionalHeader)
+		}
+
+		if h.HasProgramPacketSequenceCounter {
+			b.Write(true) // marker_bit
+			b.WriteN(h.PacketSequenceCounter, 7)
+			b.Write(true) // marker_bit
+			b.WriteN(h.MPEG1OrMPEG2ID, 1)
+			b.WriteN(h.OriginalStuffingLength, 6)
+			bytesWritten += 2
+		}
+
+		if h.HasPSTDBuffer {
+			b.WriteN(uint8(0b01), 2)
+			b.WriteN(h.PSTDBufferScale, 1)
+			b.WriteN(h.PSTDBufferSize, 13)
+			bytesWritten += 2
+		}
+
+		if h.HasExtension2 {
+			b.Write(true) // marker_bit
+			b.WriteN(uint8(len(h.Extension2Data)), 7)
+			b.Write(h.Extension2Data)
+			bytesWritten += 1 + len(h.Extension2Data)
+		}
+	}
+
+	return bytesWritten, b.Err()
+}
+
+func writeDSMTrickMode(w *astikit.BitsWriter, m *DSMTrickMode) (int, error) {
+	b := astikit.NewBitsWriterBatch(w)
+
+	b.WriteN(m.TrickModeControl, 3)
+	if m.TrickModeControl == TrickModeControlFastForward || m.TrickModeControl == TrickModeControlFastReverse {
+		b.WriteN(m.FieldID, 2)
+		b.Write(m.IntraSliceRefresh == 1) // it should be boolean
+		b.WriteN(m.FrequencyTruncation, 2)
+	} else if m.TrickModeControl == TrickModeControlFreezeFrame {
+		b.WriteN(m.FieldID, 2)
+		b.WriteN(uint8(0xff), 3) // reserved
+	} else if m.TrickModeControl == TrickModeControlSlowMotion || m.TrickModeControl == TrickModeControlSlowReverse {
+		b.WriteN(m.RepeatControl, 5)
+	} else {
+		b.WriteN(uint8(0xff), 5) // reserved
+	}
+
+	return DSMTrickModeLength, b.Err()
+}
+
+func writeESCR(w *astikit.BitsWriter, cr *ClockReference) (int, error) {
+	b := astikit.NewBitsWriterBatch(w)
+
+	b.WriteN(uint8(0xff), 2)
+	b.WriteN(uint64(cr.Base>>30), 3)
+	b.Write(true)
+	b.WriteN(uint64(cr.Base>>15), 15)
+	b.Write(true)
+	b.WriteN(uint64(cr.Base), 15)
+	b.Write(true)
+	b.WriteN(uint64(cr.Extension), 9)
+	b.Write(true)
+
+	return ESCRLength, b.Err()
+}
+
+func writePTSOrDTS(w *astikit.BitsWriter, flag uint8, cr *ClockReference) (bytesWritten int, retErr error) {
+	b := astikit.NewBitsWriterBatch(w)
+
+	b.WriteN(flag, 4)
+	b.WriteN(uint64(cr.Base>>30), 3)
+	b.Write(true)
+	b.WriteN(uint64(cr.Base>>15), 15)
+	b.Write(true)
+	b.WriteN(uint64(cr.Base), 15)
+	b.Write(true)
+
+	return PTSorDTSByteLength, b.Err()
+}
+
+func PESIsVideoStreamID(pesStreamID uint8) bool {
```

Comment:
Do you think creating a method attached to `PESHeader` instead makes more sense? Something with a name like `IsVideoStream`?

Reply:
fair enough, wil change